diff -cr src/obexd.service.in src/obexd.service.in
*** src/obexd.service.in	2011-07-16 01:26:28.000000000 +0200
--- src/obexd.service.in	2012-07-03 07:57:33.175841506 +0200
***************
*** 1,3 ****
  [D-BUS Service]
  Name=org.openobex
! Exec=@libexecdir@/obexd --nodaemon
--- 1,3 ----
  [D-BUS Service]
  Name=org.openobex
! Exec=@libexecdir@/obexd --nodaemon --noplugin=mas --symlinks --root .obex-root --root-setup /usr/bin/obex-root-setup --capability '!/usr/bin/gen-obex-capability.sh'
