commit 3f37897c8c3125cf4825f2127078c608832d730f
Author: Claudio Takahasi <claudio.takahasi@openbossa.org>
Date:   Tue Oct 27 18:24:11 2009 -0200

    First attempt to fix autostart.

diff --git a/gdbus/mainloop.c b/gdbus/mainloop.c
index a06ed22..2ff073a 100644
--- a/gdbus/mainloop.c
+++ b/gdbus/mainloop.c
@@ -272,6 +272,10 @@ DBusConnection *g_dbus_setup_bus(DBusBusType type, const char *name,
 
 	setup_dbus_with_main_loop(conn);
 
+	/* Autostart workaround: first attempt */
+	if (dbus_connection_get_dispatch_status(conn) == DBUS_DISPATCH_DATA_REMAINS)
+		g_timeout_add(DISPATCH_TIMEOUT, message_dispatch_cb, conn);
+
 	return conn;
 }
 
