diff -cr plugins/irmc.c plugins/irmc.c
*** plugins/irmc.c	Thu May 17 11:12:17 2012
--- plugins/irmc.c	Tue Jul 17 12:25:27 2012
***************
*** 226,232 ****
  	param->maxlistcount = 0; /* to count the number of vcards... */
  	param->filter = 0x200085; /* UID TEL N VERSION */
  	irmc->params = param;
! 	irmc->request = phonebook_pull("telecom/pb.vcf", irmc->params,
  					phonebook_size_result, irmc, err);
  	ret = phonebook_pull_read(irmc->request);
  	if (err)
--- 226,232 ----
  	param->maxlistcount = 0; /* to count the number of vcards... */
  	param->filter = 0x200085; /* UID TEL N VERSION */
  	irmc->params = param;
! 	irmc->request = phonebook_pull("/telecom/pb.vcf", irmc->params,
  					phonebook_size_result, irmc, err);
  	ret = phonebook_pull_read(irmc->request);
  	if (err)
***************
*** 312,318 ****
  
  	if (!g_strcmp0(name, ".vcf")) {
  		/* how can we tell if the vcard count call already finished? */
! 		irmc->request = phonebook_pull("telecom/pb.vcf", irmc->params,
  						query_result, irmc, &ret);
  		if (ret < 0) {
  			DBG("phonebook_pull failed...");
--- 312,318 ----
  
  	if (!g_strcmp0(name, ".vcf")) {
  		/* how can we tell if the vcard count call already finished? */
! 		irmc->request = phonebook_pull("/telecom/pb.vcf", irmc->params,
  						query_result, irmc, &ret);
  		if (ret < 0) {
  			DBG("phonebook_pull failed...");
