diff -cr plugins/pbap.c plugins/pbap.c
*** plugins/pbap.c	Sun Jul 22 06:16:48 2012
--- plugins/pbap.c	Sun Jul 22 06:17:10 2012
***************
*** 492,499 ****
  
  	DBG("");
  
! 	phonebook_req_finalize(pbap->obj->request);
! 	pbap->obj->request = NULL;
  
  	pbap->cache.valid = TRUE;
  
--- 492,501 ----
  
  	DBG("");
  
! 	if (pbap->obj->request) {
! 		phonebook_req_finalize(pbap->obj->request);
! 		pbap->obj->request = NULL;
! 	}
  
  	pbap->cache.valid = TRUE;
  
***************
*** 518,524 ****
  		return;
  	}
  
! 	phonebook_req_finalize(pbap->obj->request);
  	pbap->obj->request = phonebook_get_entry(pbap->folder, id,
  				pbap->params, query_result, pbap, &ret);
  	if (ret < 0)
--- 520,527 ----
  		return;
  	}
  
! 	if (pbap->obj->request)
! 		phonebook_req_finalize(pbap->obj->request);
  	pbap->obj->request = phonebook_get_entry(pbap->folder, id,
  				pbap->params, query_result, pbap, &ret);
  	if (ret < 0)
***************
*** 779,788 ****
--- 782,815 ----
  	return obj;
  }
  
+ static int vobject_close(void *object)
+ {
+ 	struct pbap_object *obj = object;
+ 
+ 	DBG("");
+ 
+ 	if (obj->session)
+ 		obj->session->obj = NULL;
+ 
+ 	if (obj->buffer)
+ 		g_string_free(obj->buffer, TRUE);
+ 
+ 	if (obj->aparams)
+ 		g_byte_array_free(obj->aparams, TRUE);
+ 
+ 	if (obj->request)
+ 		phonebook_req_finalize(obj->request);
+ 
+ 	g_free(obj);
+ 
+ 	return 0;
+ }
+ 
  static void *vobject_pull_open(const char *name, int oflag, mode_t mode,
  				void *context, size_t *size, int *err)
  {
  	struct pbap_session *pbap = context;
+ 	struct pbap_object *obj;
  	phonebook_cb cb;
  	int ret;
  	void *request;
***************
*** 810,824 ****
  	if (ret < 0)
  		goto fail;
  
  	/* reading first part of results from backend */
  	ret = phonebook_pull_read(request);
! 	if (ret < 0)
  		goto fail;
  
  	if (err)
  		*err = 0;
  
! 	return vobject_create(pbap, request);
  
  fail:
  	if (err)
--- 837,855 ----
  	if (ret < 0)
  		goto fail;
  
+ 	/* moved here from return due to synchronous backend */
+ 	obj = vobject_create(pbap, request);
  	/* reading first part of results from backend */
  	ret = phonebook_pull_read(request);
! 	if (ret < 0) {
! 		vobject_close(obj);
  		goto fail;
+ 	}
  
  	if (err)
  		*err = 0;
  
! 	return obj;
  
  fail:
  	if (err)
***************
*** 827,855 ****
  	return NULL;
  }
  
- static int vobject_close(void *object)
- {
- 	struct pbap_object *obj = object;
- 
- 	DBG("");
- 
- 	if (obj->session)
- 		obj->session->obj = NULL;
- 
- 	if (obj->buffer)
- 		g_string_free(obj->buffer, TRUE);
- 
- 	if (obj->aparams)
- 		g_byte_array_free(obj->aparams, TRUE);
- 
- 	if (obj->request)
- 		phonebook_req_finalize(obj->request);
- 
- 	g_free(obj);
- 
- 	return 0;
- }
- 
  static void *vobject_list_open(const char *name, int oflag, mode_t mode,
  				void *context, size_t *size, int *err)
  {
--- 858,863 ----
***************
*** 872,885 ****
  
  	/* PullvCardListing always get the contacts from the cache */
  
  	if (pbap->cache.valid) {
- 		obj = vobject_create(pbap, NULL);
  		ret = generate_response(pbap);
  	} else {
  		request = phonebook_create_cache(name, cache_entry_notify,
  					cache_ready_notify, pbap, &ret);
  		if (ret == 0)
! 			obj = vobject_create(pbap, request);
  	}
  	if (ret < 0)
  		goto fail;
--- 880,894 ----
  
  	/* PullvCardListing always get the contacts from the cache */
  
+ 	/* moved here due to synchronous backend */
+ 	obj = vobject_create(pbap, NULL);
  	if (pbap->cache.valid) {
  		ret = generate_response(pbap);
  	} else {
  		request = phonebook_create_cache(name, cache_entry_notify,
  					cache_ready_notify, pbap, &ret);
  		if (ret == 0)
! 			obj->request = request;
  	}
  	if (ret < 0)
  		goto fail;
***************
*** 903,908 ****
--- 912,918 ----
  					void *context, size_t *size, int *err)
  {
  	struct pbap_session *pbap = context;
+ 	struct pbap_object *obj = NULL;
  	const char *id;
  	uint32_t handle;
  	int ret;
***************
*** 921,926 ****
--- 931,938 ----
  	}
  
  	if (pbap->cache.valid == FALSE) {
+ 		/* moved here due to synchronous backend */
+ 		obj = vobject_create(pbap, NULL);
  		pbap->find_handle = handle;
  		request = phonebook_create_cache(pbap->folder,
  			cache_entry_notify, cache_entry_done, pbap, &ret);
***************
*** 933,938 ****
--- 945,952 ----
  		goto fail;
  	}
  
+ 	/* moved here due to synchronous backend */
+ 	obj = vobject_create(pbap, NULL);
  	request = phonebook_get_entry(pbap->folder, id, pbap->params,
  						query_result, pbap, &ret);
  
***************
*** 940,951 ****
  	if (ret < 0)
  		goto fail;
  
  	if (err)
  		*err = 0;
  
! 	return vobject_create(pbap, request);
  
  fail:
  	if (err)
  		*err = ret;
  
--- 954,970 ----
  	if (ret < 0)
  		goto fail;
  
+ 	obj->request = request;
+ 
  	if (err)
  		*err = 0;
  
! 	return obj;
  
  fail:
+ 	if (obj)
+ 		vobject_close(obj);
+ 
  	if (err)
  		*err = ret;
  
