diff -cr client/pbap.c client/pbap.c
*** client/pbap.c	2012-05-17 17:12:17.000000000 +0200
--- client/pbap.c	2012-08-09 17:38:03.287712242 +0200
***************
*** 233,246 ****
  
  	if (!g_ascii_strcasecmp(location, "INT") ||
  			!g_ascii_strcasecmp(location, "INTERNAL"))
! 		path = g_strdup("telecom");
  	else if (!g_ascii_strncasecmp(location, "SIM", 3)) {
  		if (strlen(location) == 3)
  			tmp = g_strdup("SIM1");
  		else
  			tmp = g_ascii_strup(location, 4);
  
! 		path = g_build_filename(tmp, "telecom", NULL);
  		g_free(tmp);
  	} else
  		return NULL;
--- 233,246 ----
  
  	if (!g_ascii_strcasecmp(location, "INT") ||
  			!g_ascii_strcasecmp(location, "INTERNAL"))
! 		path = g_strdup("/telecom");
  	else if (!g_ascii_strncasecmp(location, "SIM", 3)) {
  		if (strlen(location) == 3)
  			tmp = g_strdup("SIM1");
  		else
  			tmp = g_ascii_strup(location, 4);
  
! 		path = g_build_filename("/", tmp, "telecom", NULL);
  		g_free(tmp);
  	} else
  		return NULL;
